File: src/platform/linux/spaces_setup.cpp
#include "spaces_setup.h"
#include "spaces_nvidia_libraries.h"
#ifdef __linux__
#include <algorithm>
#include <filesystem>
#include <fstream>

namespace multiseat::spaces {
  namespace {
    using json = nlohmann::json;
    std::string either(const std::vector<std::string> &drivers) {
      std::string text;
      for (std::size_t i = 0; i < drivers.size(); ++i)
        text += (i == 0 ? "" : i + 1 == drivers.size() ? " or " : ", ") + drivers[i];
      return text;
    }
    json describe_runtime(const runtime_facts_t &r) {
      // A runtime that borrows this PC's driver is named for the graphics it
      // needs, never for a driver version: it works with whichever one is loaded.
      const auto name = !r.runtime ? std::string {} :
        r.runtime->variant == "nvidia" ? "gaming runtime for NVIDIA driver " + r.runtime->nvidia_driver :
        r.runtime->variant == "nvidia-host" ? std::string {"gaming runtime for NVIDIA graphics"} :
        std::string {"gaming runtime for AMD and Intel graphics"};
      std::string detail;
      if (r.status == "not_published") detail = "This Polaris build has no approved gaming runtime yet.";
      else if (r.code == "driver_mismatch")
        detail = "The gaming runtime in this Polaris build needs NVIDIA driver " + either(r.nvidia_drivers) + ". " +
          (r.host_nvidia_driver && !r.host_nvidia_driver->empty() ?
            "This PC runs " + *r.host_nvidia_driver + ". Install the matching driver, restart the PC, then recheck." :
            std::string {"Polaris could not read the NVIDIA driver version on this PC."});
      else if (r.code == "graphics_unsupported")
        detail = r.host_nvidia_driver ?
          std::string {"The gaming runtime in this Polaris build is for AMD and Intel graphics. This PC uses the NVIDIA driver, and no NVIDIA runtime is approved yet."} :
          "The gaming runtime in this Polaris build needs NVIDIA graphics with driver " + either(r.nvidia_drivers) +
            ". No NVIDIA driver is loaded on this PC.";
      else if (r.status == "ready") detail = "The " + name + " is downloaded and verified.";
      else if (r.code == "docker_unavailable") detail = "This PC needs the " + name + ". Polaris checks whether it is downloaded once Docker Engine is ready.";
      else if (r.status == "available") detail = "This PC needs the " + name + ". It is not downloaded yet, and the download is several gigabytes.";
      else if (r.code == "runtime_identity_mismatch")
        detail = "The gaming runtime on this PC does not match this Polaris build, so Spaces will not use it. Retry, and open Doctor & Support if it persists.";
      else detail = "Polaris could not verify the gaming runtime on this PC. Retry, and check Docker if it keeps failing.";
      json runtime {{"status", r.status}, {"code", r.code}};
      if (r.runtime) {
        runtime["id"] = r.runtime->id; runtime["variant"] = r.runtime->variant;
        runtime["nvidia_driver"] = r.runtime->nvidia_driver;
      }
      // A build without a runtime is nothing this PC can fix, so that row is
      // neutral. Downloading is the step while one fits and is not verified.
      return {{"id", "runtime"}, {"title", "Gaming runtime"},
        {"state", r.status == "ready" ? "ready" : r.status == "not_published" ? "not_configured" : "required"},
        {"detail", std::move(detail)},
        {"action", r.status == "available" ? "download_runtime" : r.status == "failed" ? "retry_runtime" : ""},
        {"doc_anchor", "#download-the-gaming-runtime"}, {"runtime", std::move(runtime)}};
    }
  }

  bool nvidia_driver_version(std::string_view value) {
    return value.size() >= 3 && value.size() <= 32 && value.front() != '.' && value.back() != '.' &&
      value.find_first_not_of("0123456789.") == std::string_view::npos &&
      value.find("..") == std::string_view::npos && value.find('.') != std::string_view::npos;
  }

  // The loaded NVIDIA kernel module names its version. Only a plain dotted
  // version is ever repeated to the reader.
  std::optional<std::string> loaded_nvidia_driver(const std::filesystem::path &module_version) {
    std::ifstream file(module_version);
    if (!file) return std::nullopt;
    std::string version(33, '\0');
    file.read(version.data(), static_cast<std::streamsize>(version.size()));
    version.resize(static_cast<std::size_t>(file.gcount()));
    while (!version.empty() && (version.back() == '\n' || version.back() == ' ')) version.pop_back();
    return nvidia_driver_version(version) ? version : std::string {};
  }

  runtime_choice_t choose_runtime(const std::vector<runtime_t> &catalog,
      const std::optional<std::string> &nvidia_driver, std::string_view profile) {
    // Only this family's entries are candidates. A build that carries a Heroic
    // runtime and no Steam one has published nothing a Steam Space can use.
    std::vector<const runtime_t *> family;
    for (const auto &runtime : catalog)
      if (runtime.profile == profile) family.push_back(&runtime);
    if (family.empty()) return {std::nullopt, "runtime_not_published"};
    if (nvidia_driver) {
      // A runtime that borrows this machine's driver is preferred, because it
      // keeps working across driver updates. It is skipped only when the
      // loaded driver is older than the one its own encoders were built for.
      bool below_minimum = false;
      for (const auto *runtime : family) {
        if (runtime->variant != "nvidia-host") continue;
        if (nvidia_driver->empty() || spaces::driver_at_least(*nvidia_driver, runtime->nvidia_minimum_driver))
          return {*runtime, {}};
        below_minimum = true;
      }
      for (const auto *runtime : family)
        if (runtime->variant == "nvidia" && runtime->nvidia_driver == *nvidia_driver) return {*runtime, {}};
      if (std::any_of(family.begin(), family.end(),
            [](const auto *runtime) { return runtime->variant == "nvidia"; }))
        return {std::nullopt, "driver_mismatch"};
      return {std::nullopt, below_minimum ? "driver_below_minimum" : "graphics_unsupported"};
    }
    for (const auto *runtime : family)
      if (runtime->variant == "default") return {*runtime, {}};
    return {std::nullopt, "graphics_unsupported"};
  }

  runtime_inspection_cache_t::runtime_inspection_cache_t(std::chrono::steady_clock::duration lifetime, now_t now) :
    lifetime_(lifetime), now_(now ? std::move(now) : now_t([] { return std::chrono::steady_clock::now(); })) {}

  runtime_image_e runtime_inspection_cache_t::remember(const std::string &reference, const std::function<runtime_image_e()> &inspect) {
    std::uint64_t generation;
    {
      std::lock_guard lock(mutex_);
      for (const auto &entry : entries_)
        if (entry.reference == reference && now_() - entry.checked < lifetime_) return entry.image;
      generation = generation_;
    }
    const auto image = inspect();
    // Timeouts and unreadable replies are asked again next time. An answer
    // that raced a download is dropped too: the download may have changed it.
    if (image != runtime_image_e::unverifiable) {
      std::lock_guard lock(mutex_);
      if (generation == generation_) {
        std::erase_if(entries_, [&](const auto &entry) { return entry.reference == reference; });
        if (entries_.size() >= 64) entries_.erase(entries_.begin());
        entries_.push_back({reference, image, now_()});
      }
    }
    return image;
  }

  void runtime_inspection_cache_t::forget() {
    std::lock_guard lock(mutex_);
    entries_.clear();
    ++generation_;
  }

  runtime_inspection_cache_t &runtime_inspection_cache() {
    static runtime_inspection_cache_t cache;
    return cache;
  }

  runtime_facts_t inspect_runtime(container::host_t &host, const std::vector<runtime_t> &catalog,
    const std::optional<std::string> &nvidia_driver, bool engine_ready, runtime_inspection_cache_t *cache,
    std::string_view profile) {
    runtime_facts_t facts;
    facts.host_nvidia_driver = nvidia_driver;
    for (const auto &runtime : catalog) {
      if (runtime.profile == profile && runtime.variant == "nvidia" &&
          std::find(facts.nvidia_drivers.begin(), facts.nvidia_drivers.end(), runtime.nvidia_driver) == facts.nvidia_drivers.end())
        facts.nvidia_drivers.push_back(runtime.nvidia_driver);
    }
    auto choice = choose_runtime(catalog, nvidia_driver, profile);
    if (!choice.runtime) {
      facts.status = choice.code == "runtime_not_published" ? "not_published" : "unsupported";
      facts.code = std::move(choice.code);
      return facts;
    }
    facts.runtime = std::move(choice.runtime);
    if (!engine_ready) {
      facts.status = "available"; facts.code = "docker_unavailable";
      return facts;
    }
    const auto &runtime = *facts.runtime;
    const auto inspect = [&] {
      auto argv = container::command_prefix({});
      argv.insert(argv.end(), {"image", "inspect", runtime.reference()});
      const auto result = host.run(argv, std::chrono::seconds(5), 65536);
      if (result.timed_out || result.output_truncated) return runtime_image_e::unverifiable;
      if (result.exit_status == 0) {
        if (verified_runtime_image(runtime, result.output)) return runtime_image_e::verified;
        // Docker described one image under the pinned reference, and it differs.
        const auto images = json::parse(result.output, nullptr, false);
        return images.is_array() && images.size() == 1 ? runtime_image_e::mismatch : runtime_image_e::unverifiable;
      }
      // Docker lists what it found, an empty list, and exits 1 when nothing
      // has this reference. The engine answered moments ago.
      std::string_view found = result.output;
      while (!found.empty() && (found.back() == '\n' || found.back() == ' ')) found.remove_suffix(1);
      return result.exit_status == 1 && found == "[]" ? runtime_image_e::absent : runtime_image_e::unverifiable;
    };
    const auto image = cache ? cache->remember(runtime.reference(), inspect) : inspect();
    facts.status = image == runtime_image_e::verified ? "ready" : image == runtime_image_e::absent ? "available" : "failed";
    facts.code = image == runtime_image_e::verified ? "runtime_ready" : image == runtime_image_e::absent ? "not_downloaded" :
      image == runtime_image_e::mismatch ? "runtime_identity_mismatch" : "inspection_failed";
    return facts;
  }

  bool docker_access_pending(const std::optional<container::group_membership_t> &docker, std::uint64_t effective_gid,
    const std::optional<std::vector<std::uint64_t>> &groups) {
    return docker && docker->member && groups && effective_gid != docker->gid &&
      std::find(groups->begin(), groups->end(), docker->gid) == groups->end();
  }

  namespace {
    /** What this distribution calls the 32 bit half of the NVIDIA driver. */
    std::string thirty_two_bit_driver_package(std::string_view distribution) {
      if (distribution == "fedora" || distribution == "bazzite") return "xorg-x11-drv-nvidia-libs.i686";
      if (distribution == "arch" || distribution == "cachyos") return "lib32-nvidia-utils";
      if (distribution == "ubuntu") return "the libnvidia-gl package for your driver branch, i386 variant";
      return {};
    }
  }  // namespace

  nlohmann::json describe_setup(const setup_facts_t &f) {
    using json = nlohmann::json;
    json checks = json::array();
    // Every check names the guide section that fixes it, so a failing row is
    // never a dead end whatever the console decides to render.
    auto add = [&](const char *id, const char *title, bool ready, const char *detail,
                   const char *action, const char *anchor) {
      checks.push_back({{"id", id}, {"title", title}, {"state", ready ? "ready" : "required"},
        {"detail", detail}, {"action", action}, {"doc_anchor", anchor}});
    };
    add("docker", "Docker Engine", f.docker_cli && f.runc,
      f.docker_cli && f.runc ? "Docker and its container runtime are installed." :
      "Install Docker Engine on this PC to run separate gaming spaces.", "install_docker", "#prepare-docker-from-spaces");
    const bool engine = f.daemon_replied && f.daemon_linux && !f.daemon_rootless && f.daemon_runc;
    const bool pending = !engine && !f.daemon_replied && f.docker_access_pending;
    add("docker_access", "Polaris access to Docker", engine,
      engine ? "This Polaris service can reach the local Docker Engine." :
      pending ? "Polaris was given access to Docker after it started, and a running Polaris keeps the access it started with. Restart this PC, then recheck." :
      !f.daemon_replied ? "Start Docker and allow the Polaris service account to use it." :
      f.daemon_rootless ? "This version requires the system Docker Engine. Rootless Docker is not supported for Spaces yet." :
      "Spaces needs a local Linux Docker Engine with runc.", "docker_access", "#prepare-docker-from-spaces");
    // Starting Docker and joining its group is what an administrator can approve from Polaris.
    // A daemon that answers but is rootless or not runc, or access that only needs a restart, is not.
    if (!engine && !pending && !f.daemon_replied && f.docker_cli && f.runc && !f.immutable_host)
      checks.back()["host_action"] = "docker_access";
    add("identity", "Gaming runtime account", f.uid == 1000 && f.gid == 1000,
      f.uid == 1000 && f.gid == 1000 ? "The current runtime supports this service account." :
      "The current preview runtime does not yet support this service account. Do not change your Linux user ID.",
      "", "#gaming-runtime-account");
    add("input", "Controller and input access", f.input_access,
      f.input_access ? "Polaris can create virtual input devices. Controls are checked again when a space starts." :
      "Complete Polaris host setup so streamed controls can reach a space.", "host_setup", "#controller-access");
    add("gpu", "Graphics device access", f.gpu_access,
      f.gpu_access ? "A graphics device is accessible. Hardware encoding is checked when the space starts." :
      "Polaris cannot access a graphics device. Check the driver and host permissions.", "host_setup", "#graphics-access");
    if (f.runtime.host_nvidia_driver && !f.runtime.host_nvidia_driver->empty()) {
      const bool ready = f.driver_libraries.empty();
      const auto package = f.driver_libraries_package.empty() ?
        std::string {"this distribution's 32 bit NVIDIA driver package"} : f.driver_libraries_package;
      const std::string detail =
        ready ? "This PC's NVIDIA driver files are ready for a space to borrow." :
        f.driver_libraries == "driver_libraries_32bit_missing" ?
          "Install " + package + ". A space can still start, but 32 bit games, which is most games under Proton, render nothing." :
        f.driver_libraries == "driver_below_minimum" ?
          std::string {"This PC's NVIDIA driver is older than the gaming runtime supports. Update the driver."} :
        f.driver_libraries == "driver_libraries_untrusted" ?
          std::string {"Some NVIDIA driver files on this PC are not owned by root, so Polaris will not pass them to a space."} :
        f.driver_libraries == "driver_configuration_missing" ?
          std::string {"This PC's Vulkan or EGL description for NVIDIA could not be read."} :
          std::string {"This PC's NVIDIA driver files are not where the driver package puts them."};
      add("nvidia_libraries", "NVIDIA driver files", ready, detail.c_str(), "", "#nvidia-driver-files");
    }
    add("security", "Spaces security support", f.security.ready, f.security.detail.c_str(), f.security.code.c_str(),
      "#prepare-spaces-security-support");
    // The packaged helper installs or updates the policies; a host that is not enforcing, or whose
    // package is incomplete, needs something an administrator prompt cannot give.
    if (!f.security.ready && f.security.code == "install_selinux" && !f.immutable_host)
      checks.back()["host_action"] = "security_install";
    // Not a host prerequisite: first-Space setup downloads the runtime too.
    checks.push_back(describe_runtime(f.runtime));
    checks.push_back({{"id", "spaces"}, {"title", "Spaces configuration"},
      {"state", f.controller_available ? "ready" : f.controller_enabled ? "required" : "not_configured"},
      {"detail", f.controller_available ? "The configured Spaces controller is available." :
        f.controller_enabled ? "The configured Spaces controller is unavailable. Review the setup diagnostics." :
        "No Space has been configured on this host yet."},
      {"action", "configure_spaces"}, {"doc_anchor", "#prepare-your-first-space"}});
    return {{"version", 2}, {"distribution", f.distribution},
      {"immutable_host", f.immutable_host}, {"service_uid", f.uid},
      {"host_prerequisites_ready", f.docker_cli && f.runc && engine && f.uid == 1000 && f.gid == 1000 && f.input_access && f.gpu_access && f.security.ready},
      {"configured", f.controller_enabled}, {"available", f.controller_available},
      {"checks", std::move(checks)}};
  }

  nlohmann::json inspect_setup(container::host_t &host, bool enabled, bool available, const std::optional<security_facts_t> &security) {
    setup_facts_t f;
    f.security = describe_security(security ? *security : inspect_security(host));
    f.uid = host.effective_uid(); f.gid = host.effective_gid();
    f.controller_enabled = enabled; f.controller_available = available;
    std::error_code error;
    f.immutable_host = std::filesystem::exists("/run/ostree-booted", error);
    // If the host type cannot be determined, do not offer mutable package steps.
    if (error) f.immutable_host = true;
    std::ifstream release("/etc/os-release");
    std::string line;
    // Read a literal ID, never source distribution files as shell code.
    for (unsigned lines = 0; lines < 64 && std::getline(release, line); ++lines) {
      if (line.starts_with("ID=")) {
        auto id = line.substr(3);
        if (id.size() >= 2 && id.front() == '"' && id.back() == '"') id = id.substr(1, id.size() - 2);
        if (id == "fedora" || id == "arch" || id == "cachyos" || id == "ubuntu" || id == "bazzite" || id == "steamos")
          f.distribution = id;
      }
    }
    f.docker_cli = host.trusted_runtime_file("/usr/bin/docker");
    f.runc = host.trusted_runtime_file("/usr/bin/runc");
    if (f.docker_cli && f.runc) {
      auto argv = container::command_prefix({});
      argv.insert(argv.end(), {"info", "--format={{json .}}"});
      auto result = host.run(argv, std::chrono::seconds(3), 65536);
      if (!result.timed_out && !result.output_truncated && result.exit_status == 0) {
        try {
          const auto info = nlohmann::json::parse(result.output);
          if (info.is_object() && info.contains("SecurityOptions") && info["SecurityOptions"].is_array()) {
            f.daemon_replied = true; f.daemon_linux = info.value("OSType", "") == "linux";
            for (const auto &option : info["SecurityOptions"]) {
              if (!option.is_string()) { f.daemon_replied = false; break; }
              if (option.get<std::string>().starts_with("name=rootless")) f.daemon_rootless = true;
            }
            auto runtime = info.at("Runtimes").at("runc").value("path", "");
            f.daemon_runc = runtime == "runc" || runtime == "/usr/bin/runc";
          }
        } catch (...) { f.daemon_replied = false; }
      }
      // Docker refused this process. If the account already joined the group, only a restart helps.
      if (!f.daemon_replied)
        f.docker_access_pending = docker_access_pending(host.group_membership("docker"), host.effective_gid(), host.supplementary_groups());
    }
    f.input_access = host.read_write_character_device("/dev/uinput").has_value() &&
      host.read_write_character_device("/dev/uhid").has_value();
    // Limit the read-only device probe. Seat admission still selects and
    // verifies exact physical devices; this does not allocate or open a GPU.
    for (unsigned i = 128; i < 192; ++i) {
      if (host.read_write_character_device("/dev/dri/renderD" + std::to_string(i))) {
        f.gpu_access = true; break;
      }
    }
    static const std::vector<runtime_t> unpublished;
    const auto &catalog = trusted_runtimes();
    const auto loaded = loaded_nvidia_driver();
    f.runtime = inspect_runtime(host, catalog ? *catalog : unpublished, loaded,
      f.docker_cli && f.runc && f.daemon_replied && f.daemon_linux && !f.daemon_rootless && f.daemon_runc,
      &runtime_inspection_cache());
    // Only relevant with an NVIDIA driver loaded: whether this PC can lend its
    // own driver files to a space. Reading files, never running anything.
    if (loaded && !loaded->empty()) {
      if (const auto &contract = trusted_nvidia_contract()) {
        const auto facts = resolve_host_driver_libraries(*contract, host, *loaded, contract->minimum_driver);
        f.driver_libraries = facts.code;
        f.driver_libraries_package = thirty_two_bit_driver_package(f.distribution);
      }
    }
    return describe_setup(f);
  }
}
#endif

File: src_assets/common/assets/web/spaces-setup.js
// Commands are fixed application text, never shell instructions returned by an API.
export const hostChecks = ['docker', 'docker_access', 'identity', 'input', 'gpu', 'security']
// A host sends these only when they apply: the runtime check arrived in 1.4.9,
// and the NVIDIA driver files check exists only while an NVIDIA driver is
// loaded. Neither one decides host_prerequisites_ready.
const optionalCheckIds = ['runtime', 'nvidia_libraries']
const checkIds = [...hostChecks, ...optionalCheckIds, 'spaces']
const runtimeStatuses = ['not_published', 'unsupported', 'available', 'ready', 'failed']

// The gaming runtime check, since Polaris 1.4.9. Its status picks the row; its
// state keeps the grammar every check shares, and only a build without a
// runtime is neutral. A runtime is named only when one fits this PC.
export function validRuntimeCheck(item) {
  const runtime = item?.runtime
  if (!runtime || typeof runtime !== 'object' || !runtimeStatuses.includes(runtime.status) ||
      typeof runtime.code !== 'string' || !/^[a-z0-9_]{1,64}$/.test(runtime.code)) return false
  if (item.state !== (runtime.status === 'ready' ? 'ready' : runtime.status === 'not_published' ? 'not_configured' : 'required')) return false
  if (!['available', 'ready', 'failed'].includes(runtime.status)) return runtime.id === undefined
  // Only a runtime that carries NVIDIA userspace of its own names a driver
  // version. One that borrows this PC's works with whichever driver is loaded,
  // so it names none, exactly like the runtime for AMD and Intel graphics.
  if (!['default', 'nvidia', 'nvidia-host'].includes(runtime.variant)) return false
  return typeof runtime.id === 'string' && /^[a-z0-9][a-z0-9-]{0,63}$/.test(runtime.id) &&
    typeof runtime.nvidia_driver === 'string' && (runtime.variant === 'nvidia'
      ? runtime.nvidia_driver.length <= 32 && /^[0-9]+(?:\.[0-9]+)+$/.test(runtime.nvidia_driver)
      : runtime.nvidia_driver === '')
}

export function validSetup(value) {
  if (!value || value.version !== 2 || typeof value.distribution !== 'string' ||
      typeof value.immutable_host !== 'boolean' ||
      !Number.isSafeInteger(value.service_uid) || value.service_uid < 0 ||
      !['host_prerequisites_ready', 'configured', 'available'].every(key => typeof value[key] === 'boolean') ||
      !Array.isArray(value.checks)) return false
  // Every check a host does send must be one this console knows, and every
  // check that always applies must be there.
  const seen = new Set()
  for (const item of value.checks) {
    if (!item || !checkIds.includes(item.id) || seen.has(item.id) ||
        !['ready', 'required', 'not_configured'].includes(item.state) ||
        typeof item.title !== 'string' || typeof item.detail !== 'string' || typeof item.action !== 'string' ||
        (item.doc_anchor !== undefined && !/^#[a-z0-9-]{1,64}$/.test(item.doc_anchor))) return false
    if (!['spaces', 'runtime'].includes(item.id) && item.state === 'not_configured') return false
    if (item.id === 'runtime' && !validRuntimeCheck(item)) return false
    seen.add(item.id)
  }
  if (!checkIds.every(id => optionalCheckIds.includes(id) || seen.has(id))) return false
  if (!seen.has('spaces')) return false
  return value.host_prerequisites_ready === hostChecks.every(id => value.checks.find(item => item.id === id).state === 'ready') &&
    (!value.available || value.configured) &&
    (value.checks.find(item => item.id === 'spaces').state === 'ready') === value.available
}

export const installGuides = {
  fedora: {
    name: 'Fedora',
    url: 'https://docs.docker.com/engine/install/fedora/',
    steps: [
      { title: 'Add the Docker package repository', command: 'sudo dnf config-manager addrepo --from-repofile https://download.docker.com/linux/fedora/docker-ce.repo' },
      { title: 'Install Docker Engine', command: 'sudo dnf install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin' },
    ],
  },
  arch: {
    name: 'Arch Linux',
    url: 'https://wiki.archlinux.org/title/Docker',
    steps: [{ title: 'Update Arch and install Docker', command: 'sudo pacman -Syu docker' }],
  },
  ubuntu: {
    name: 'Ubuntu',
    url: 'https://docs.docker.com/engine/install/ubuntu/',
    steps: [
      { title: 'Install the repository prerequisites', command: 'sudo apt update\nsudo apt install ca-certificates curl' },
      { title: 'Add the Docker signing key', command: 'sudo install -m 0755 -d /etc/apt/keyrings\nsudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc\nsudo chmod a+r /etc/apt/keyrings/docker.asc' },
      { title: 'Add the Docker package repository', command: 'sudo tee /etc/apt/sources.list.d/docker.sources <<EOF\nTypes: deb\nURIs: https://download.docker.com/linux/ubuntu\nSuites: $(. /etc/os-release && echo "${UBUNTU_CODENAME:-$VERSION_CODENAME}")\nComponents: stable\nArchitectures: $(dpkg --print-architecture)\nSigned-By: /etc/apt/keyrings/docker.asc\nEOF\nsudo apt update' },
      { title: 'Install Docker Engine', command: 'sudo apt install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin' },
    ],
  },
}
export const startDocker = 'sudo systemctl enable --now docker'
export function dockerAccessCommand(uid) {
  return Number.isSafeInteger(uid) && uid > 0 && uid <= 2147483647
    ? 'sudo usermod -aG docker -- "$(id -nu -- ' + uid + ')"' : ''
}
export function installGuide(setup) {
  if (!setup || setup.immutable_host) return null
  return Object.hasOwn(installGuides, setup.distribution) ? installGuides[setup.distribution] : null
}

export const installSpacesSecurity = 'sudo -H /usr/bin/polaris-spaces-setup install'
export const fedoraSecurityPackages = 'sudo dnf install selinux-policy-devel container-selinux make'

// The guide on papi-ux.com. A check names its own section (doc_anchor) since
// Polaris 1.4.9; older hosts get the section the console has always linked.
export const docsUrl = 'https://papi-ux.com/docs/spaces/'
const legacyAnchors = {
  docker: '#prepare-docker-from-spaces', docker_access: '#prepare-docker-from-spaces', identity: '#gaming-runtime-account',
  input: '#controller-access', gpu: '#graphics-access', security: '#prepare-spaces-security-support',
  runtime: '#download-the-gaming-runtime', spaces: '#prepare-your-first-space',
}
export function guideHref(check) {
  const anchor = typeof check?.doc_anchor === 'string' && /^#[a-z0-9-]{1,64}$/.test(check.doc_anchor)
    ? check.doc_anchor : legacyAnchors[check?.id] || ''
  return docsUrl + anchor
}

// The terminal steps a failing check needs, in order, from this file's fixed
// text only. Package installs are withheld on an immutable host; starting the
// daemon and granting the service account access apply everywhere.
export function setupSteps(setup, check) {
  if (!setup || !check || check.state === 'ready') return []
  const guide = installGuide(setup)
  if (check.id === 'docker') {
    return [...(guide?.steps || []), { title: 'Start Docker', command: startDocker }]
  }
  if (check.id === 'docker_access') {
    const access = dockerAccessCommand(setup.service_uid)
    return [{ title: 'Start Docker', command: startDocker },
      ...(access ? [{ title: 'Let the Polaris service account use Docker', command: access }] : [])]
  }
  if (check.id === 'security') {
    if (setup.immutable_host) return []
    return [...(setup.distribution === 'fedora' ? [{ title: 'Install the policy tools', command: fedoraSecurityPackages }] : []),
      { title: 'Run the helper from the native package, with Polaris stopped', command: installSpacesSecurity }]
  }
  return []
}
